README.md
# caclj

An [elementary cellular automata](https://en.wikipedia.org/wiki/Elementary_cellular_automaton) with clojure.

## Usage

```bash
$ lein run -- --help
```

```bash
An elementary cellular automata.

Usage: lein run [options]

Options:
  -r, --rule RULE    30  Rule number
  -t, --times TIMES  16  The number ob the generations
  -h, --help
```

## Examples

```bash
$ lein run -r 30
```

```bash
                o
               ooo
              oo  o
             oo oooo
            oo  o   o
           oo oooo ooo
          oo  o    o  o
         oo oooo  oooooo
        oo  o   ooo     o
       oo oooo oo  o   ooo
      oo  o    o oooo oo  o
     oo oooo  oo o    o oooo
    oo  o   ooo  oo  oo o   o
   oo oooo oo  ooo ooo  oo ooo
  oo  o    o ooo   o  ooo  o  o
 oo oooo  oo o  o ooooo  ooooooo
```

```bash
$ lein run -r 110 -t 16
```

```bash
                o
               oo
              ooo
             oo o
            ooooo
           oo   o
          ooo  oo
         oo o ooo
        ooooooo o
       oo     ooo
      ooo    oo o
     oo o   ooooo
    ooooo  oo   o
   oo   o ooo  oo
  ooo  oooo o ooo
 oo o oo  ooooo o
```

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-06-12
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-06-12
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/caclj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/caclj/compare/0.1.0...0.1.1


test/caclj/core_test.clj
(ns caclj.core-test
  (:require [clojure.test :refer :all]
            [caclj.core :refer :all]))


(deftest test-init-rule
  (is (= (init-rule 0)
         [false false false false false false false false]))
  (is (= (init-rule 1)
         [true false false false false false false false]))
  (is (= (init-rule 30)
         [false true true true true false false false]))
  (is (= (init-rule 255)
         [true true true true true true true true]))
  (is (thrown? java.lang.AssertionError (init-rule -1)))
  (is (thrown? java.lang.AssertionError (init-rule 256))))


(deftest test-next-gen
  (is (= (next-gen (init-rule 30) [true])
         [true true true])))


test/caclj/cellular_automata_test.clj
(ns caclj.cellular-automata-test
  (:require [clojure.test :refer :all]
            [caclj.cellular-automata :refer :all]))

(deftest init-rule-test
  (is (= (init-rule 0)
         [false false false false false false false false]))
  (is (= (init-rule 1)
         [true false false false false false false false]))
  (is (= (init-rule 30)
         [false true true true true false false false]))
  (is (= (init-rule 255)
         [true true true true true true true true])))

(deftest next-gen-test
  (is (= (next-gen (init-rule 30) [true])
         [true true true])))


src/caclj/cellular_automata.clj
(ns caclj.cellular-automata
  (:require [clojure.string :as string]
            [clojure.spec.alpha :as s]
            [clojure.spec.test.alpha :as stest]))

(def RULE-SEQUENCE-SIZE 8)
(def INITIAL-CELLS [true])

(s/def :caclj/rule (s/coll-of boolean? :kind vector? :count RULE-SEQUENCE-SIZE))
(s/def :caclj/cells (s/coll-of boolean? :kind vector? :min-count 1))

(defn init-rule
  "Returns a boolean sequence that expresses the rule number."
  [number]
  (let [bin-str (Integer/toBinaryString number)
        pad (repeat (- RULE-SEQUENCE-SIZE (count bin-str)) \0)]
    (mapv #(if (= \1 %) true false) (reverse (concat pad bin-str)))))

(s/fdef init-rule
        :args (s/and (s/cat :number int?) #(<= 0 (:number %) 255))
        :ret :caclj/rule)
;(stest/instrument `init-rule)
;(init-rule -1)
;(init-rule 256)
;(stest/check `init-rule)

(defn next-gen
  "Returns the next generations."
  [rule cells]
  (let [len (count cells)
        fun #(let [_p (dec %)
                   _n (inc %)
                   l (if (and (<= 0 _p) (< _p len) (get cells (mod _p len))) 4 0)
                   c (if (and (<= 0 %) (< % len) (get cells %)) 2 0)
                   r (if (and (<= 0 _n) (< _n len) (get cells (mod _n len))) 1 0)]
               (get rule (+ l c r)))]
    (mapv fun (range -1 (+ len 1)))))

(s/fdef next-gen
        :args (s/cat :rule :caclj/rule :cells :caclj/cells)
        :ret :caclj/cells)

(defn print-cells
  "Prints cells nicely."
  [cells t]
  (let [pad (string/join "" (repeat t " "))
        s (string/join "" (map #(if (true? %) "o" " ") cells))]
    (println (str pad s))))

(s/fdef print-cells
        :args (s/and (s/cat :cells :caclj/cells :t int?) #(< 0 (:t %)))
        :ret nil)


src/caclj/core.clj
(ns caclj.core
  (:require [clojure.string :as string]
            [clojure.tools.cli :refer [parse-opts]]
            [caclj.cellular-automata :as ca])
  (gen-class))


(def cli-options
  [["-r" "--rule RULE" "Rule number"
    :default 30
    :parse-fn #(Integer/parseInt %)
    :validate [#(<= 0 % 255) "Must be a number between 0 and 255"]]
   ["-t" "--times TIMES" "The number ob the generations"
    :default 16
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 %) "Must be a positive number"]]
   ["-h" "--help"]])


(defn usage [options-summary]
  (->> ["An elementary cellular automata."
        ""
        "Usage: lein run [options]"
        ""
        "Options:"
        options-summary]
       (string/join \newline)))


(defn error-msg [errors]
  (str "The following errors occurred while parsing your command:\n\n"
       (string/join \newline errors)))


(defn validate-args
  "Validate command line arguments. "
  [args]
  (let [{:keys [options arguments errors summary]} (parse-opts args cli-options)]
    (cond
      (:help options) ; help => exit OK with usage summary
      {:exit-message (usage summary) :ok? true}
      errors ; errors => exit with description of errors
      {:exit-message (error-msg errors)}
      :else
      {:options options})))


(defn exit [status msg]
  (println msg)
  (System/exit status))


(defn -main
  [& args]
  (let [{:keys [options exit-message ok?]} (validate-args args)]
    (if exit-message
      (exit (if ok? 0 1) exit-message)
      (let [rule (ca/init-rule (:rule options))]
        (loop [cells ca/INITIAL-CELLS
               t (:times options)]
          (ca/print-cells cells t)
          (if (> (dec t) 0)
            (recur (ca/next-gen rule cells) (dec t))))))))
